File: Mantle/MTLValueTransformer.m
//
//  MTLValueTransformer.m
//  Mantle
//
//  Created by Justin Spahr-Summers on 2012-09-11.
//  Copyright (c) 2012 GitHub. All rights reserved.
//

#import "MTLValueTransformer.h"

/// A block that represents a transformation.
///
/// value   - The value to transform.
/// success - The block must set this parameter to indicate whether the
///           transformation was successful.
///           MTLValueTransformer will always call this block with *success
///           initialized to YES.
/// error   - If not NULL, this may be set to an error that occurs during
///           transforming the value.
///
/// Returns the result of the transformation, which may be nil.
typedef id _Nullable (^MTLAnyValueTransformerBlock)(_Nullable __kindof id value, BOOL *_Nonnull success, NSError *_Nullable *_Nullable error);

//
// Any MTLValueTransformer supporting reverse transformation. Necessary because
// +allowsReverseTransformation is a class method.
//
@interface MTLReversibleValueTransformer : MTLValueTransformer
@end

@interface MTLValueTransformer ()

@property (nonatomic, copy, readonly) MTLAnyValueTransformerBlock forwardBlock;
@property (nonatomic, copy, readonly) MTLAnyValueTransformerBlock reverseBlock;

@end

@implementation MTLValueTransformer

#pragma mark Lifecycle

+ (instancetype)transformerUsingForwardBlock:(MTLAnyValueTransformerBlock)forwardBlock {
	return [[self alloc] initWithForwardBlock:forwardBlock reverseBlock:nil];
}

+ (instancetype)transformerUsingReversibleBlock:(MTLAnyValueTransformerBlock)reversibleBlock {
	return [self transformerUsingForwardBlock:reversibleBlock reverseBlock:reversibleBlock];
}

+ (instancetype)transformerUsingForwardBlock:(MTLAnyValueTransformerBlock)forwardBlock reverseBlock:(MTLAnyValueTransformerBlock)reverseBlock {
	return [[MTLReversibleValueTransformer alloc] initWithForwardBlock:forwardBlock reverseBlock:reverseBlock];
}

- (instancetype)initWithForwardBlock:(MTLAnyValueTransformerBlock)forwardBlock reverseBlock:(MTLAnyValueTransformerBlock)reverseBlock {
	NSParameterAssert(forwardBlock != nil);

	self = [super init];
	if (self == nil) return nil;

	_forwardBlock = [forwardBlock copy];
	_reverseBlock = [reverseBlock copy];

	return self;
}

#pragma mark NSValueTransformer

+ (BOOL)allowsReverseTransformation {
	return NO;
}

+ (Class)transformedValueClass {
	return NSObject.class;
}

- (id)transformedValue:(id)value {
	NSError *error = nil;
	BOOL success = YES;

	return self.forwardBlock(value, &success, &error);
}

- (id)transformedValue:(id)value success:(BOOL *)outerSuccess error:(NSError **)outerError {
	NSError *error = nil;
	BOOL success = YES;

	id transformedValue = self.forwardBlock(value, &success, &error);

	if (outerSuccess != NULL) *outerSuccess = success;
	if (outerError != NULL) *outerError = error;

	return transformedValue;
}

@end

@implementation MTLReversibleValueTransformer

#pragma mark Lifecycle

- (id)initWithForwardBlock:(MTLAnyValueTransformerBlock)forwardBlock reverseBlock:(MTLAnyValueTransformerBlock)reverseBlock {
	NSParameterAssert(reverseBlock != nil);
	return [super initWithForwardBlock:forwardBlock reverseBlock:reverseBlock];
}

#pragma mark NSValueTransformer

+ (BOOL)allowsReverseTransformation {
	return YES;
}

- (id)reverseTransformedValue:(id)value {
	NSError *error = nil;
	BOOL success = YES;

	return self.reverseBlock(value, &success, &error);
}

- (id)reverseTransformedValue:(id)value success:(BOOL *)outerSuccess error:(NSError **)outerError {
	NSError *error = nil;
	BOOL success = YES;

	id transformedValue = self.reverseBlock(value, &success, &error);

	if (outerSuccess != NULL) *outerSuccess = success;
	if (outerError != NULL) *outerError = error;

	return transformedValue;
}

@end


@implementation MTLValueTransformer (Deprecated)

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-implementations"

+ (instancetype)transformerWithBlock:(id (^)(id))transformationBlock {
	return [self transformerUsingForwardBlock:^(id value, BOOL *success, NSError **error) {
		return transformationBlock(value);
	}];
}

+ (instancetype)reversibleTransformerWithBlock:(id (^)(id))transformationBlock {
	return [self transformerUsingReversibleBlock:^(id value, BOOL *success, NSError **error) {
		return transformationBlock(value);
	}];
}

+ (instancetype)reversibleTransformerWithForwardBlock:(id (^)(id))forwardBlock reverseBlock:(id (^)(id))reverseBlock {
	return [self
		transformerUsingForwardBlock:^(id value, BOOL *success, NSError **error) {
			return forwardBlock(value);
		}
		reverseBlock:^(id value, BOOL *success, NSError **error) {
			return reverseBlock(value);
		}];
}

#pragma clang diagnostic pop

@end

File: Mantle/MTLValueTransformer.swift
//
//  MTLValueTransformer.swift
//  Mantle
//
//  Created by Zachary Waldowski on 1/21/16.
//  Copyright © 2016. Some rights reserved. Licensed under MIT.
//

import Mantle.MTLValueTransformer

private func objCApplyTransform<In, Out>(transform: In throws -> Out, value: AnyObject?, success: UnsafeMutablePointer<ObjCBool>, error: NSErrorPointer) -> AnyObject? {
	do {
		if case nil = value { return nil }
		guard let input = value as? In else {
			throw MTLValueTransformerError.InvalidInput(value ?? NSNull())
		}
		return try transform(input) as? AnyObject
	} catch let localError as NSError {
		if (success != nil) { success.memory = false }
		if (error != nil) { error.memory = localError }
		return nil
	}
}

/// A value transformer supporting function-based transformation.
public final class MTLValueTransformer<In, Out>: NSValueTransformer {

	/// The forward transform. May be invoked manually.
	public let transform: In throws -> Out

	/// Creates a forward transformer which transforms values using the given
	/// function. Reverse transformations will not be allowed.
	public init(transform: In throws -> Out) {
		self.transform = transform
		super.init()
	}

	// MARK: - NSValueTransformer

	/// Always `false`.
	/// - seealso: MTLReversibleValueTransformer
	public override class func allowsReverseTransformation() -> Bool {
		return false
	}

	public override class func transformedValueClass() -> AnyClass {
		return NSObject.self
	}

    /// Transforms a value.
	public override func transformedValue(value: AnyObject?) -> AnyObject? {
		return objCApplyTransform(transform, value: value, success: nil, error: nil)
	}

	// MARK: - MTLTransformerErrorHandling

    /// Transforms a value, returning any error that occurred.
    ///
    /// - parameter value: The value to transform.
    /// - parameter success: If not NULL, on return will indicate whether the
    ///   whether the transformation was successful.
    /// - parameter error: If not NULL, on return may be set to an error that
    ///   occured while transforming the value.
	/// - returns: The result of the transformation. Users should inspect
    ///   `success` to decide how to proceed with the result.
	public override func transformedValue(value: AnyObject?, success: UnsafeMutablePointer<ObjCBool>, error: NSErrorPointer) -> AnyObject? {
		return objCApplyTransform(transform, value: value, success: success, error: error)
	}

}

/// A value transformer supporting reversible function-based transformation.
public final class MTLReversibleValueTransformer<In, Out>: NSValueTransformer {

	/// The forward transform. May be invoked manually.
	public let transform: In throws -> Out

	/// The reverse transform. May be invoked manually.
	public let reverseTransform: Out throws -> In

	/// Creates a transformer which forward and reverse transforms values using
	/// the given functions.
	public init(forward transform: In throws -> Out, reverse reverseTransform: Out throws -> In) {
		self.transform = transform
		self.reverseTransform = reverseTransform
		super.init()
	}

	// MARK: - NSValueTransformer

	/// Always `true`.
	/// - seealso: MTLValueTransformer
	public override class func allowsReverseTransformation() -> Bool {
		return true
	}

	public override class func transformedValueClass() -> AnyClass {
		return NSObject.self
	}

	/// Transforms a value.
	public override func transformedValue(value: AnyObject?) -> AnyObject? {
		return objCApplyTransform(transform, value: value, success: nil, error: nil)
	}

	/// Reverse-transforms a value.
	public override func reverseTransformedValue(value: AnyObject?) -> AnyObject? {
		return objCApplyTransform(reverseTransform, value: value, success: nil, error: nil)
	}

	// MARK: - MTLTransformerErrorHandling

    /// Transforms a value, returning any error that occurred.
    ///
    /// - parameter value: The value to transform.
    /// - parameter success: If not NULL, on return will indicate whether the
    ///   whether the transformation was successful.
    /// - parameter error: If not NULL, on return may be set to an error that
    ///   occured while transforming the value.
	/// - returns: The result of the transformation. Users should inspect
    ///   `success` to decide how to proceed with the result.
	public override func transformedValue(value: AnyObject?, success: UnsafeMutablePointer<ObjCBool>, error: NSErrorPointer) -> AnyObject? {
		return objCApplyTransform(transform, value: value, success: success, error: error)
	}

	/// Reverse-transforms a value, returning any error that occurred.
	///
    /// - parameter value: The value to transform.
    /// - parameter success: If not NULL, on return will indicate whether the
    ///   whether the transformation was successful.
    /// - parameter error: If not NULL, on return may be set to an error that
    ///   occured while transforming the value.
	/// - returns: The result of the transformation. Users should inspect
    ///   `success` to decide how to proceed with the result.
	public override func reverseTransformedValue(value: AnyObject?, success: UnsafeMutablePointer<ObjCBool>, error: NSErrorPointer) -> AnyObject? {
		return objCApplyTransform(reverseTransform, value: value, success: success, error: error)
	}

}

extension MTLReversibleValueTransformer {

	/// Returns a transformer by flipping the direction of `self`, such that
	/// a forward transformation becomes a reverse transformation, and vice-versa.
	public func inverted() -> MTLReversibleValueTransformer<Out, In> {
		return .init(forward: reverseTransform, reverse: transform)
	}

}
